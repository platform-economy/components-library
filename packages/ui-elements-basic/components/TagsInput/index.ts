export * from './TagsInput';
export * from './TagsList';
export * from './OptionsList';
export * from './TagsContextWrapper';
