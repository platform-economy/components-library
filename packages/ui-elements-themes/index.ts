export * from './themes/light';
export * from './themes/dark';
export * from './themes/test';
export * from './themes/_common/paletteDark';
export * from './themes/_common/paletteLight';
export * from './themes/_common/methods';
export * from './themes/_common/fonts';
export * from './themes/_common/withTestTheme';
