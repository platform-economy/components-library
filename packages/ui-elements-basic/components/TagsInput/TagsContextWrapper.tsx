import * as React from 'react';
import _ from 'lodash';

import { TagType } from '@relayr/ui-elements-basic/components/Tag';

export interface TagsContextProps{
  selectedTags?: TagType[];
  sourceTags?: TagType[];
  optionsTags?: TagType[];
  selectTag?: (tag:TagType) => void;
  removeTag?: (tag:TagType) => void;
  selectAllTags?: () => void;
  deselectAllTags?: () => void;
  onValueChange?: (value: string | string[] | number, ...args: unknown[]) => void;
  inputValue?: string | string[] | number;
  invalid?: boolean;
  noTagsMsg?:string;
  handleKeyDown?: (event: React.KeyboardEvent) => void;
  enableCustomTags?: boolean;
  inputValidator?: (inputValue:string | string[] | number) => boolean;
}

export interface TagsContextInterface{
  selectedTags?: TagType[];
  sourceTags?: TagType[];
  optionsTags?: TagType[];
  selectTag?: (tag:TagType) => void;
  removeTag?: (tag:TagType) => void;
  selectAllTags?: () => void;
  deselectAllTags?: () => void;
  onValueChange?: (value: string | string[] | number, ...args: unknown[]) => void;
  inputValue?: string | string[] | number;
  invalid?: boolean;
  noTagsMsg?:string;
  handleKeyDown?: (event: React.KeyboardEvent) => void;
}

interface TagsContextState {
  selectedTags?: TagType[];
  sourceTags?: TagType[];
  optionsTags?: TagType[];
  selectTag?: (tag:TagType) => void;
  removeTag?: (tag:TagType) => void;
  selectAllTags?: () => void;
  deselectAllTags?: () => void;
  onValueChange?: (value: string | string[] | number, ...args: unknown[]) => void;
  inputValue?: string | string[] | number;
  invalid?: boolean;
  noTagsMsg?:string;
  handleKeyDown?: (event: React.KeyboardEvent) => void;
}

export const TagsContext =
  React.createContext<TagsContextInterface|null>(null);

export class TagsContextWrapper extends  React.Component<TagsContextProps, TagsContextState>{
  private lastTagRemoveByKeyboard: number;
  private delayBetweenDeleteTabByKeyboard:number = 500;
  constructor(props:TagsContextProps, ...args:unknown[]) {
    super(props, ...args);
    this.state = {
      selectedTags: this.props.selectedTags || [],
      sourceTags: this.props.sourceTags,
      optionsTags: this.props.optionsTags ? this.props.optionsTags : this.props.sourceTags || [],
      selectTag: this.selectTag,
      removeTag: this.removeTag,
      selectAllTags: this.selectAllTags,
      deselectAllTags: this.deselectAllTags,
      onValueChange: this.onValueChange,
      inputValue: '',
      invalid: false,
      noTagsMsg: this.props.noTagsMsg,
      handleKeyDown: this.handleKeyDown,
    };
  }

  selectTag = (tag:TagType) => {
    if (this.state.selectedTags
      && this.state.selectedTags.filter(e => e.id === tag.id).length <= 0) {
      const options = _.reject(this.state.optionsTags, e => e.id === tag.id);
      this.setState({
        selectedTags: [...this.state.selectedTags, tag],
        optionsTags: [...options],
      });
    }
    this.clearRemovingState();
  }

  removeTag = (tag:TagType) => {
    if (this.state.optionsTags
      && this.state.optionsTags.filter(e => e.id === tag.id).length <= 0) {
      let selected = _.reject(this.state.selectedTags, e => e.id === tag.id);
      selected = selected.map(tag => ({ ...tag, removing: false }));
      if (tag.custom) {
        this.setState({
          selectedTags: [...selected],
          optionsTags: [...this.state.optionsTags],
        });
      } else {
        this.setState({
          selectedTags: [...selected],
          optionsTags: [...this.state.optionsTags, tag],
        });
      }

    }
  }

  selectAllTags = () => {
    this.setState({
      selectedTags: _.uniqBy([...this.state.selectedTags, ...this.state.optionsTags], t => t.id),
      optionsTags: [],
    });
    this.clearRemovingState();
  }

  deselectAllTags = () => {
    const selectedTagsToRemove = _.filter(this.state.selectedTags, e => e.custom !== true);
    this.setState({
      selectedTags: [],
      optionsTags: _.uniqBy(
        [...this.props.sourceTags, ...selectedTagsToRemove, ...this.state.optionsTags],
        t => t.id,
        ),
      inputValue: '',
      invalid: this.validateInput(''),
    });
  }

  validateInput = (value: string | string[] | number) => {
    // returns true means invalid | false means valid
    return (value
    && this.props.inputValidator
    && !this.props.inputValidator(value)) || false;
  }

  onValueChange = (value: string | string[] | number) => {
    this.setState({
      inputValue: value,
      invalid: this.validateInput(value),
    });
  }

  clearRemovingState = () => {
    this.setState(prevState => ({
      selectedTags: prevState.selectedTags && prevState.selectedTags.map(
        el => ({ ...el, removing: false }),
      ),
      optionsTags: prevState.optionsTags && prevState.optionsTags.map(
        el => ({ ...el, removing: false }),
      ),
    }));
  }

  clearInputValue = () => {
    this.setState({
      inputValue: '',
    });
  }

  addCustomTag = () => {
    const selectedTags = this.state.selectedTags;
    const customTag: TagType = {
      id: `${this.state.inputValue
        && this.state.inputValue.toString()}${selectedTags && selectedTags.length}`,
      title: this.state.inputValue && this.state.inputValue.toString() || '',
      custom: true,
    };
    if (this.props.inputValidator && this.state.inputValue) {
      if (!this.validateInput(this.state.inputValue)) {
        this.setState({
          selectedTags: [...this.state.selectedTags, customTag],
        });
        this.clearInputValue();
      }
    } else {
      this.setState({
        selectedTags: [...this.state.selectedTags, customTag],
      });
      this.clearInputValue();
    }
  }

  handleKeyDown = (e: React.KeyboardEvent<HTMLInputElement>) => {
    if (
      e.keyCode === 8
      && this.state.selectedTags
      && this.state.selectedTags.length
      && (this.state.inputValue === undefined || this.state.inputValue === '')
    ) {
      if (this.state.selectedTags[this.state.selectedTags.length - 1].removing !== true) {
        this.setState(prevState => ({
          selectedTags: prevState.selectedTags && prevState.selectedTags.map(
            (el, i) => (prevState.selectedTags
              && i === prevState.selectedTags.length - 1) ? { ...el, removing: true } : el,
          ),
        }));
        this.state.selectedTags[this.state.selectedTags.length - 1].removing = true;
      } else if (
        this.lastTagRemoveByKeyboard == null ||
        this.lastTagRemoveByKeyboard + this.delayBetweenDeleteTabByKeyboard < Date.now()
      ) {
        this.removeTag(this.state.selectedTags[this.state.selectedTags.length - 1]);
        this.clearRemovingState();
        this.lastTagRemoveByKeyboard = Date.now();
      }
    } else if (
      e.keyCode === 13
      && this.props.enableCustomTags
      && this.state.inputValue
      && this.state.inputValue.toString().length > 0
      ) {
      this.addCustomTag();
    } else if (
      e.keyCode === 9
      && this.props.enableCustomTags
      && this.state.inputValue
      && this.state.inputValue.toString().length > 0
      ) {
      e.preventDefault();
      this.addCustomTag();
    }
  }

  static getDerivedStateFromProps(newProps: TagsContextProps, state: TagsContextState) {
    if (newProps.sourceTags !== state.sourceTags) {
      const newOptionsTags:TagType[] = [];
      _.each(newProps.sourceTags, (tag:TagType) => {
        if (_.findIndex(state.selectedTags, (st:TagType) => st.id === tag.id) === -1) {
          newOptionsTags.push(tag);
        }
      });
      return {
        sourceTags: newProps.sourceTags,
        optionsTags: newOptionsTags,
      };
    }
    return null;
  }

  render() {
    return(
      <TagsContext.Provider value={this.state}>
        {this.props.children}
      </TagsContext.Provider>
    );
  }
}
