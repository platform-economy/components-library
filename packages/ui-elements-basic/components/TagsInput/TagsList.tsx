import * as React from 'react';
import styled from 'styled-components';
import { Spinner } from '@relayr/ui-elements-skeletons';
// import classNames from 'classnames';

import { Tag, TagType } from '@relayr/ui-elements-basic/components/Tag';
import { Button } from '@relayr/ui-elements-basic/components/Button';

export type TagsListProps = {
  options?: TagType[];
  clearIcon?: React.ReactNode;
  selectAllTags?: () => void;
  noTagsMsg?:string;
  loading?:boolean;
};

const StyledTagsList = styled.ul<TagsListProps>`
  display: flex;
  flex-direction: row;
  min-height: 35px;
  width: 100%;
  box-shadow: 0 1px 3px 0 rgba(0, 0, 0, 0.23);
  background-color: ${({ theme }) => theme.palette.surface};
  color: ${({ theme }) => theme.palette.text};
  cursor: text;
  border: none;
  margin: 0;
  padding: .2em;
  box-sizing: border-box;
  align-items: center;
  overflow: auto;
  flex-wrap: wrap;
  & li{
    margin-left: .5em;
    margin-top: 2px;
    list-style-type: none;
  }
  ${Button.Tertiary}{
    margin-left: auto;
    padding: 5px;
    min-height: 25px;
  }
  span.noTagsMsg{
    color: ${({ theme }) => theme.palette.text};
    margin-left: 1em;
    font-size: .75em;
  }
  & .loadingSpinner{
    margin: 2px;
  }
`;

const StyledLi = styled.li`
  list-style-type: none;
`;

export class TagsList extends React.Component<TagsListProps> {
  render() {
    const { options, clearIcon, selectAllTags, noTagsMsg } = this.props;
    return (
      <StyledTagsList {...this.props}>
        {options && options.map((tag: TagType) => {
          return (
            <StyledLi key={tag.id}>
              <Tag
                tagId={tag.id}
                title={tag.title}
                selected={false}
                removeIcon={clearIcon}
                custom={tag.custom}
                content={tag.content}
              >
                <span>{tag.content || tag.title}</span>
              </Tag>
            </StyledLi>
          );
        })
        }
        <span className="noTagsMsg">{options && options.length ? null : noTagsMsg}</span>
        {this.props.loading ? <Spinner className="loadingSpinner" /> : null}
        {options && options.length
          ? <Button.Tertiary onClick={selectAllTags}>Select All</Button.Tertiary>
          : null
        }
      </StyledTagsList>
    );
  }
}
