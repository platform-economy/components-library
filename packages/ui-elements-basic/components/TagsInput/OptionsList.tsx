import * as React from 'react';
import styled from 'styled-components';
// import classNames from 'classnames';

import { TagType } from '@relayr/ui-elements-basic/components/Tag';
import { TagsContext, TagsContextProps } from './TagsContextWrapper';
import { Spinner } from '@relayr/ui-elements-skeletons';

export type OptionsListProps = {
  options?: TagType[];
  clearIcon?: React.ReactNode;
  selectAllTags?: () => void;
  noTagsMsg?:string;
  loading?:boolean;
};

export type OptionProps = {
  tagId?: string;
  title?: string;
  selected?: boolean;
  selectHandler?: () => void;
  removeHandler?: () => void;
};

const StyledOptionsList = styled.ul<OptionsListProps>`
  display: flex;
  flex-direction: column;
  width: 100%;
  min-height: 1em;
  box-shadow: 0 1px 3px 0 rgba(0, 0, 0, 0.23);
  background-color: ${({ theme }) => theme.palette.surface};
  color: ${({ theme }) => theme.palette.text};
  cursor: pointer;
  border: none;
  margin: 0;
  padding: 0;
  box-sizing: border-box;
  align-items: flex-start;
  overflow: auto;
  flex-wrap: wrap;
  & li{
    width: 100%;
    min-height: 1em;
    padding: 0;
    line-height: 2em;
    box-sizing: border-box;
  }
  span.noTagsMsg{
    color: ${({ theme }) => theme.palette.text};
    font-size: .875em;
    margin-left: .75em;
    line-height: 2em;
  }
`;

const StyledOption = styled.li`
  min-height: 1.25em;
  color: ${props => props.theme.palette.text};
  line-height: 1.25em;
  list-style-type: none;
  font-weight: normal;
  font-size: .875em;
  text-align: center;
  border-left: 2px solid transparent;
  :hover {
    background: ${({ theme }) => theme.palette.midground};
    border-left: 2px solid ${({ theme }) => theme.palette.active};;
  };

  :active {
    color: ${({ theme }) => theme.palette.text_highlight};
    background: ${({ theme }) => theme.palette.active_lowered};
  };
  & span{
    padding-left: 1em;
  }
  & button{
    background: none;
    outline: none;
    border: none;
    padding: 0;
    color: ${({ theme }) => theme.palette.text};
  }
  &:focus-within{
    background-color:${({ theme }) => theme.palette.active_hovered};
  }
  & .loadingSpinner{
    display: inline;
    padding: 2px;
  }
`;

export class OptionsList extends React.Component<OptionsListProps> {
  render() {
    const { options, noTagsMsg } = this.props;
    return (
      <TagsContext.Consumer>
        {(context: TagsContextProps) => (
          <StyledOptionsList {...this.props}>
            {options && options.map((tag: TagType) => {
              const selectTag = () => {
                context.selectTag
                  && context.selectTag({ title: tag.title, id: tag.id, content: tag.content });
              };
              return (
                <StyledOption
                  key={tag.id}
                  onClick={selectTag}
                >
                  <button>{tag.content || tag.title}</button>
                </StyledOption>
              );
            })}
            <span className="noTagsMsg">{options && options.length ? null : noTagsMsg}</span>
            <span>{this.props.loading ? <Spinner className="loadingSpinner" /> : null}</span>
          </StyledOptionsList>
        )}
      </TagsContext.Consumer>);
  }
}
