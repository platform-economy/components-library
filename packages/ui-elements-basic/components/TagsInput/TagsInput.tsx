import * as React from 'react';
import styled from 'styled-components';
import classNames from 'classnames';
import _ from 'lodash';

import { Tag, TagType } from '@relayr/ui-elements-basic/components/Tag';
import { Dropdown, DropdownStated } from '@relayr/ui-elements-basic/components/Dropdown';
import { TagsList, OptionsList } from './index';

import { TagsContextWrapper, TagsContext, TagsContextInterface } from './TagsContextWrapper';

export type TagsInputProps = {
  placeholder?: string;
  clearIcon?: React.ReactNode;
  selectedTags?: TagType[];
  optionsTags?: TagType[];
  sourceTags?: TagType[];
  onBlur?: React.FocusEventHandler;
  onChange?: React.FormEventHandler;
  onClick?: React.MouseEventHandler;
  onFocus?: React.FocusEventHandler;
  clearTags?: () => void;
  componentsLeft?: React.ReactNode;
  componentsRight?: React.ReactNode;
  tagsContext?: TagsContextInterface;
  onValueChange?: (value: string | string[] | number, ...args: unknown[]) => void;
  inputValue?: string | string[] | number;
  invalid?: boolean;
  listShown?: boolean;
  optionsList?:boolean;
  loading?:boolean;
  noTagsMsg?:string;
  onTagsQueryChange?:(query?:string) => void;
  onKeyDown?: (event: React.KeyboardEvent) => void;
  enableCustomTags?: boolean;
  inputValidator?: (inputValue:string | string[] | number) => boolean;
  className?: string;
  throttle?: number;
};

export type TagsInputState = {
  hasFocus?: boolean;
};

const StyledTagsInput = styled.ul<TagsInputProps>`
  display: flex;
  flex-direction: row;
  min-height: 35px;
  box-shadow: 0 1px 3px 0 rgba(0, 0, 0, 0.23);
  background-color: ${({ theme }) => theme.palette.foreground};
  color: ${({ theme }) => theme.palette.text};
  cursor: text;
  border: none;
  margin: 0;
  padding-left: 0;
  padding-right: 25px;
  padding-bottom: 2px;
  padding-top: 2px;
  box-sizing: border-box;
  border-bottom: 2px solid transparent;
  align-items: center;
  overflow: auto;
  flex-wrap: wrap;
  position: relative;
  user-select: none;
  width: 100%;

  &:invalid{
    border-bottom: 2px solid ${props => props.theme.palette.negative};
    outline: transparent;
  }

  & li{
    margin-left: .5em;
    margin-top: 2px;
    list-style-type: none;
  }

  & > input {
    background: transparent;
    color: ${({ theme }) => theme.palette.text};
    border: none;
    flex: 1 1 auto;
    padding: 0 .5em;
    margin-top: 2px;
    min-height: 26px;
  }

  & > input:focus{
    border: none;
    outline: none;
  }

  &:invalid{
    border-bottom: 2px solid ${props => props.theme.palette.negative};
    outline: transparent;
  }
  @media screen and (min-width: 414px){
    & > input::-webkit-input-placeholder {
    font-size: .875em;
    }
    & > input::-moz-placeholder {
      font-size: .875em;
    }
    & > input:-ms-input-placeholder {
      font-size: .875em;
    }
    & > input::placeholder {
      font-size: .875em;
    }
  }

  &.hasFocus{
    border-bottom: 2px solid ${props => props.theme.palette.active};
  }

  &.invalid{
    border-bottom: 2px solid ${props => props.theme.palette.negative};
  }

  & > .clearBtn{
    font-size: 1em;
    position: absolute;
    top: 8px;
    right: 10px;
    transition: all .3s ease;
    color: ${props => props.theme.palette.text_inverted};
    pointer-events:none;
    opacity: 0;
    transform: rotate(0deg);
  }

  & > .clearBtn.clearVisible{
    cursor: pointer;
    top: 8px;
    right: 10px;
    opacity: 1;
    pointer-events:auto;
    transform: rotate(90deg);
  }
  & .clearBtn:hover{
    color: ${props => props.theme.palette.text}
  }
  & .clearBtn:active{
    color: ${props => props.theme.palette.active_lowered}
  }

  & div.DropdownContent{
    box-shadow: 0 1px 3px 0 rgba(0, 0, 0, 0.23);
  }
`;

const ClearBtn = styled.button`
  background: none;
  border: none;
  outline: none;
  padding: 0;
  :focus{
    color: ${props => props.theme.palette.active_hovered}
  }
`;

const StyledLi = styled.li`
  list-style-type: none;
`;

class TagsInputPure extends React.Component<TagsInputProps, TagsInputState> {
  private emitAutocompleteUpdateThrottled:() => void;
  private TagsInputRef: React.RefObject<HTMLInputElement>;
  static contextType = TagsContext;
  state = {
    hasFocus: false,
  };

  constructor(props:TagsInputProps) {
    super(props);
    this.TagsInputRef = React.createRef();
    this.emitAutocompleteUpdateThrottled =
    _.throttle(this.emitAutocompleteUpdate, this.props.throttle || 500);
  }

  handleChange = (event:React.ChangeEvent<HTMLInputElement>) => {
    this.context.onValueChange && this.context.onValueChange(event.target.value);
    this.props.onChange && this.props.onChange(event);
    this.emitAutocompleteUpdateThrottled();
  }

  handleClick = (event: React.MouseEvent) => {
    this.TagsInputRef.current!.focus();
    this.setState({ hasFocus: true });
    event.preventDefault();
  }

  handleFocus = (event:React.FocusEvent<HTMLInputElement>) => {
    if (this.props.onFocus) {
      this.props.onFocus(event);
    }
    this.setState({ hasFocus: true });
  }

  handleBlur = (event:React.FocusEvent<HTMLInputElement>) => {
    if (this.props.onBlur) {
      this.props.onBlur(event);
    }
    this.setState({ hasFocus: false });
  }

  clearInputInternal = () => {
    this.context.deselectAllTags && this.context.deselectAllTags();
    this.props.onTagsQueryChange && this.props.onTagsQueryChange('');
    this.props.clearTags ? this.props.clearTags() : null;
  }

  emitAutocompleteUpdate = () => {
    if (this.props.onTagsQueryChange != null && this.TagsInputRef.current) {
      this.props.onTagsQueryChange(
        this.TagsInputRef.current.value ? this.TagsInputRef.current.value.toString() : undefined);
    }
  }

  render() {
    const {
      placeholder,
      clearIcon,
      componentsLeft,
      componentsRight,
      optionsList,
      loading,
    } = this.props;
    const {
      inputValue,
      optionsTags,
      selectAllTags,
      noTagsMsg,
      invalid,
      handleKeyDown,
    } = this.context;
    const selectedTags = this.props.selectedTags || this.context.selectedTags;
    const clearVisible = selectedTags
    && selectedTags.length || inputValue;
    const clearBtn = clearIcon
      ? (
      <ClearBtn
        onClick={this.clearInputInternal}
        className={classNames('clearBtn', { clearVisible })}
      >{clearIcon}
      </ClearBtn>
      )
    : null ;

    const tagsOptionsList = () => {
      if (!optionsList) {
        return (
              <TagsList
                options={optionsTags}
                clearIcon={clearIcon}
                selectAllTags={selectAllTags}
                noTagsMsg={noTagsMsg}
                loading={loading}
              />
        );
      } return (
            <OptionsList
              options={optionsTags}
              clearIcon={clearIcon}
              selectAllTags={selectAllTags}
              noTagsMsg={noTagsMsg}
              loading={loading}
            />
      );
    };

    return (
        <DropdownStated
          {...this.props}
          dropOnFocus={true}
          style={{ width: '100%' }}
        >
          <Dropdown.Cover disableClick={true}>
            <StyledTagsInput
              className={
                classNames(
                  'TagsInput',
                  {
                    invalid,
                    hasFocus: this.state.hasFocus,
                  },
                )
              }
              onClick={this.handleClick}
            >
              {componentsLeft}
              {selectedTags && selectedTags.map((tag:TagType) => {
                return (
                  <StyledLi key={tag.id}>
                    <Tag
                      tagId={tag.id}
                      title={tag.title}
                      selected={true}
                      removeIcon={clearIcon || true}
                      removing={tag.removing}
                      custom={tag.custom}
                      content={tag.content}
                    >
                      <span>{tag.content || tag.title}</span>
                    </Tag>
                  </StyledLi>
                );
              })
              }
              <input
                ref={this.TagsInputRef}
                type="text"
                placeholder={placeholder}
                onFocus={this.handleFocus}
                onBlur={this.handleBlur}
                onChange={this.handleChange}
                value={inputValue}
                onKeyDown={handleKeyDown}
              />

              {clearBtn}
              {componentsRight}
            </StyledTagsInput>
          </Dropdown.Cover>
          <Dropdown.Content alignment="justify">
              {tagsOptionsList()}
          </Dropdown.Content>
        </DropdownStated>
    );
  }
}

export class TagsInput extends  React.Component<TagsInputProps> {
  render() {
    const { sourceTags, noTagsMsg, enableCustomTags, inputValidator } = this.props;
    return (
      <TagsContextWrapper
        sourceTags={sourceTags}
        noTagsMsg={noTagsMsg}
        enableCustomTags={enableCustomTags}
        inputValidator={inputValidator}
      >
        <TagsInputPure
          className="TagsInput"
          {...this.props}
        />
      </TagsContextWrapper>
    );
  }
}
