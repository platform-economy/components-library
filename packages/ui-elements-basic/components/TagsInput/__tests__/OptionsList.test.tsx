import * as React from 'react';
import * as enzyme from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import * as TestRenderer from 'react-test-renderer';
import 'jest-styled-components';

import { withTestTheme } from '@relayr/ui-elements-themes';

// Component
import { OptionsList, TagsContext } from '../index';

enzyme.configure({ adapter: new Adapter() });

const sourceTags = [{ id:'1', title: 'test' }];

import { TagsContextWrapper } from '../../TagsInput';

const WrappingComponent = (props: { children: React.ReactElement }) => (
  withTestTheme(props.children)
);

describe('ui-elements-basic', () => {
  describe('Components', () => {
    describe('TagsInput', () => {
      it('should render tagsList and optionsList with empty props', () => {
        const tree = TestRenderer
            .create(withTestTheme(
                <OptionsList options={[]} />))
            .toJSON();
        expect(tree).toMatchSnapshot();
      });
      it('should render tagsList and optionsList with props', () => {
        const tree = TestRenderer
            .create(withTestTheme(
                <OptionsList options={sourceTags} />))
            .toJSON();
        expect(tree).toMatchSnapshot();
      });
      it('should render tagsList and optionsList with props', () => {
        const tree = TestRenderer
            .create(withTestTheme(
                <OptionsList options={sourceTags} loading={true}/>))
            .toJSON();
        expect(tree).toMatchSnapshot();
      });
      it('should render tagsList and optionsList with empty props and loading', () => {
        const tree = TestRenderer
            .create(withTestTheme(
                <OptionsList options={[]} loading={true}/>))
            .toJSON();
        expect(tree).toMatchSnapshot();
      });
      it('Should handle remove tag onClick event correctly', () => {
        const context = { selectTag: jest.fn() };
        const tree = enzyme.mount(
          withTestTheme(
            <TagsContext.Provider value={context}>
              <OptionsList options={sourceTags} />,
             </TagsContext.Provider>,
          ),
          {
            childContextTypes: context,
            wrappingComponent: WrappingComponent,
          },
        );
        tree.find('button').simulate('click');
        expect(context.selectTag).toBeCalledWith({
          id: sourceTags[0].id,
          title: sourceTags[0].title,
        });
      });
      it('Should handle remove tag onClick event correctly', () => {
        const tree = enzyme.mount(
          withTestTheme(
            <TagsContextWrapper>
              <OptionsList options={sourceTags}/>
            </TagsContextWrapper>,
          ),
          { wrappingComponent: WrappingComponent },
        );
        tree.find('button').simulate('click');
      });
    });
  });
});
