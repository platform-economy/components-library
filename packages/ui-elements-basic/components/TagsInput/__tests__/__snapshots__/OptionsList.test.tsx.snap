// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ui-elements-basic Components TagsInput should render tagsList and optionsList with empty props 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  width: 100%;
  min-height: 1em;
  box-shadow: 0 1px 3px 0 rgba(0,0,0,0.23);
  background-color: #surface;
  color: #text;
  cursor: pointer;
  border: none;
  margin: 0;
  padding: 0;
  box-sizing: border-box;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  overflow: auto;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
}

.c0 li {
  width: 100%;
  min-height: 1em;
  padding: 0;
  line-height: 2em;
  box-sizing: border-box;
}

.c0 span.noTagsMsg {
  color: #text;
  font-size: .875em;
  margin-left: .75em;
  line-height: 2em;
}

<ul
  className="c0"
>
  <span
    className="noTagsMsg"
  />
  <span />
</ul>
`;

exports[`ui-elements-basic Components TagsInput should render tagsList and optionsList with empty props and loading 1`] = `
.c1 {
  height: 1em;
  width: 1em;
  border-radius: 50%;
  border: .2em solid #text_inverted;
  border-top: .2em solid #active;
  -webkit-animation: hmHMaM 1s infinite linear;
  animation: hmHMaM 1s infinite linear;
  box-sizing: border-box;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  width: 100%;
  min-height: 1em;
  box-shadow: 0 1px 3px 0 rgba(0,0,0,0.23);
  background-color: #surface;
  color: #text;
  cursor: pointer;
  border: none;
  margin: 0;
  padding: 0;
  box-sizing: border-box;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  overflow: auto;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
}

.c0 li {
  width: 100%;
  min-height: 1em;
  padding: 0;
  line-height: 2em;
  box-sizing: border-box;
}

.c0 span.noTagsMsg {
  color: #text;
  font-size: .875em;
  margin-left: .75em;
  line-height: 2em;
}

<ul
  className="c0"
>
  <span
    className="noTagsMsg"
  />
  <span>
    <div
      className="loadingSpinner c1"
    />
  </span>
</ul>
`;

exports[`ui-elements-basic Components TagsInput should render tagsList and optionsList with props 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  width: 100%;
  min-height: 1em;
  box-shadow: 0 1px 3px 0 rgba(0,0,0,0.23);
  background-color: #surface;
  color: #text;
  cursor: pointer;
  border: none;
  margin: 0;
  padding: 0;
  box-sizing: border-box;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  overflow: auto;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
}

.c0 li {
  width: 100%;
  min-height: 1em;
  padding: 0;
  line-height: 2em;
  box-sizing: border-box;
}

.c0 span.noTagsMsg {
  color: #text;
  font-size: .875em;
  margin-left: .75em;
  line-height: 2em;
}

.c1 {
  min-height: 1.25em;
  color: #text;
  line-height: 1.25em;
  list-style-type: none;
  font-weight: normal;
  font-size: .875em;
  text-align: center;
  border-left: 2px solid transparent;
}

.c1:hover {
  background: #midground;
  border-left: 2px solid #active;
}

.c1:active {
  color: #text_highlight;
  background: #active_lowered;
}

.c1 span {
  padding-left: 1em;
}

.c1 button {
  background: none;
  outline: none;
  border: none;
  padding: 0;
  color: #text;
}

.c1:focus-within {
  background-color: #active_hovered;
}

.c1 .loadingSpinner {
  display: inline;
  padding: 2px;
}

<ul
  className="c0"
>
  <li
    className="c1"
    onClick={[Function]}
  >
    <button>
      test
    </button>
  </li>
  <span
    className="noTagsMsg"
  />
  <span />
</ul>
`;

exports[`ui-elements-basic Components TagsInput should render tagsList and optionsList with props 2`] = `
.c2 {
  height: 1em;
  width: 1em;
  border-radius: 50%;
  border: .2em solid #text_inverted;
  border-top: .2em solid #active;
  -webkit-animation: hmHMaM 1s infinite linear;
  animation: hmHMaM 1s infinite linear;
  box-sizing: border-box;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  width: 100%;
  min-height: 1em;
  box-shadow: 0 1px 3px 0 rgba(0,0,0,0.23);
  background-color: #surface;
  color: #text;
  cursor: pointer;
  border: none;
  margin: 0;
  padding: 0;
  box-sizing: border-box;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  overflow: auto;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
}

.c0 li {
  width: 100%;
  min-height: 1em;
  padding: 0;
  line-height: 2em;
  box-sizing: border-box;
}

.c0 span.noTagsMsg {
  color: #text;
  font-size: .875em;
  margin-left: .75em;
  line-height: 2em;
}

.c1 {
  min-height: 1.25em;
  color: #text;
  line-height: 1.25em;
  list-style-type: none;
  font-weight: normal;
  font-size: .875em;
  text-align: center;
  border-left: 2px solid transparent;
}

.c1:hover {
  background: #midground;
  border-left: 2px solid #active;
}

.c1:active {
  color: #text_highlight;
  background: #active_lowered;
}

.c1 span {
  padding-left: 1em;
}

.c1 button {
  background: none;
  outline: none;
  border: none;
  padding: 0;
  color: #text;
}

.c1:focus-within {
  background-color: #active_hovered;
}

.c1 .loadingSpinner {
  display: inline;
  padding: 2px;
}

<ul
  className="c0"
>
  <li
    className="c1"
    onClick={[Function]}
  >
    <button>
      test
    </button>
  </li>
  <span
    className="noTagsMsg"
  />
  <span>
    <div
      className="loadingSpinner c2"
    />
  </span>
</ul>
`;
